File: frontend/.intlayer/types/crossAlerts.ts
/* eslint-disable */
export default {
  "key": "crossAlerts",
  "content": {
    "nodeType": "translation",
    "translation": {
      "en": {
        "title": "Cross-Alert Correlations",
        "subtitle": "Correlations between anomaly and drift alerts",
        "sections": {
          "relatedDriftAlerts": "Related Drift Alerts",
          "relatedAnomalyAlerts": "Related Anomaly Alerts",
          "correlations": "Alert Correlations",
          "autoTriggerConfig": "Auto-Trigger Configuration",
          "recentEvents": "Recent Events"
        },
        "strength": {
          "strong": "Strong",
          "moderate": "Moderate",
          "weak": "Weak",
          "none": "None"
        },
        "alertType": {
          "anomaly": "Anomaly",
          "drift": "Drift"
        },
        "labels": {
          "sourceId": "Source",
          "confidence": "Confidence",
          "timeDelta": "Time Difference",
          "commonColumns": "Common Columns",
          "suggestedAction": "Suggested Action",
          "anomalyRate": "Anomaly Rate",
          "anomalyCount": "Anomaly Count",
          "driftPercentage": "Drift Percentage",
          "driftedColumns": "Drifted Columns",
          "createdAt": "Created",
          "correlationStrength": "Correlation"
        },
        "config": {
          "enabled": "Auto-Trigger Enabled",
          "triggerDriftOnAnomaly": "Trigger drift check on anomaly spike",
          "triggerAnomalyOnDrift": "Trigger anomaly check on drift detection",
          "notifyOnCorrelation": "Notify when correlation detected",
          "cooldownSeconds": "Cooldown (seconds)",
          "thresholds": {
            "title": "Trigger Thresholds",
            "anomalyRateThreshold": "Anomaly Rate Threshold",
            "anomalyCountThreshold": "Anomaly Count Threshold",
            "driftPercentageThreshold": "Drift Percentage Threshold",
            "driftColumnsThreshold": "Drifted Columns Threshold"
          }
        },
        "events": {
          "triggerType": {
            "anomaly_to_drift": "Anomaly -> Drift",
            "drift_to_anomaly": "Drift -> Anomaly",
            "bidirectional": "Bidirectional"
          },
          "status": {
            "pending": "Pending",
            "running": "Running",
            "completed": "Completed",
            "failed": "Failed",
            "skipped": "Skipped"
          },
          "correlationFound": "Correlation Found",
          "noCorrelation": "No Correlation"
        },
        "stats": {
          "totalCorrelations": "Total Correlations",
          "strongCorrelations": "Strong",
          "recentCorrelations": "Last 24h",
          "recentTriggers": "Auto-Triggers (24h)"
        },
        "actions": {
          "viewDetails": "View Details",
          "goToAnomaly": "Go to Anomaly Detection",
          "goToDrift": "Go to Drift Monitoring",
          "refresh": "Refresh",
          "configure": "Configure",
          "triggerNow": "Trigger Now",
          "saveConfig": "Save Configuration"
        },
        "messages": {
          "configSaved": "Configuration saved",
          "triggerStarted": "Trigger started",
          "triggerCompleted": "Trigger completed",
          "triggerFailed": "Trigger failed",
          "noCorrelationsFound": "No correlations found",
          "loadingCorrelations": "Loading correlations...",
          "errorLoadingCorrelations": "Failed to load correlations"
        },
        "empty": {
          "noCorrelations": "No correlations yet",
          "noCorrelationsDesc": "Correlations will appear when both anomaly and drift alerts occur for the same source",
          "noRelatedAlerts": "No related alerts",
          "noRelatedAlertsDesc": "No correlated alerts found for this source in the selected time window",
          "noEvents": "No auto-trigger events",
          "noEventsDesc": "Auto-trigger events will appear here when conditions are met"
        },
        "time": {
          "seconds": "seconds",
          "minutes": "minutes",
          "hours": "hours",
          "ago": "ago",
          "apart": "apart"
        }
      },
      "ko": {
        "title": "교차 알림 상관관계",
        "subtitle": "이상 탐지와 드리프트 알림 간의 상관관계",
        "sections": {
          "relatedDriftAlerts": "관련 드리프트 알림",
          "relatedAnomalyAlerts": "관련 이상 탐지 알림",
          "correlations": "알림 상관관계",
          "autoTriggerConfig": "자동 트리거 설정",
          "recentEvents": "최근 이벤트"
        },
        "strength": {
          "strong": "강함",
          "moderate": "중간",
          "weak": "약함",
          "none": "없음"
        },
        "alertType": {
          "anomaly": "이상",
          "drift": "드리프트"
        },
        "labels": {
          "sourceId": "소스",
          "confidence": "신뢰도",
          "timeDelta": "시간 차이",
          "commonColumns": "공통 컬럼",
          "suggestedAction": "권장 조치",
          "anomalyRate": "이상 비율",
          "anomalyCount": "이상 수",
          "driftPercentage": "드리프트 비율",
          "driftedColumns": "드리프트 컬럼",
          "createdAt": "생성일",
          "correlationStrength": "상관관계"
        },
        "config": {
          "enabled": "자동 트리거 활성화",
          "triggerDriftOnAnomaly": "이상 급증 시 드리프트 검사 트리거",
          "triggerAnomalyOnDrift": "드리프트 감지 시 이상 검사 트리거",
          "notifyOnCorrelation": "상관관계 감지 시 알림",
          "cooldownSeconds": "쿨다운 (초)",
          "thresholds": {
            "title": "트리거 임계값",
            "anomalyRateThreshold": "이상 비율 임계값",
            "anomalyCountThreshold": "이상 수 임계값",
            "driftPercentageThreshold": "드리프트 비율 임계값",
            "driftColumnsThreshold": "드리프트 컬럼 수 임계값"
          }
        },
        "events": {
          "triggerType": {
            "anomaly_to_drift": "이상 -> 드리프트",
            "drift_to_anomaly": "드리프트 -> 이상",
            "bidirectional": "양방향"
          },
          "status": {
            "pending": "대기 중",
            "running": "실행 중",
            "completed": "완료",
            "failed": "실패",
            "skipped": "건너뜀"
          },
          "correlationFound": "상관관계 발견",
          "noCorrelation": "상관관계 없음"
        },
        "stats": {
          "totalCorrelations": "전체 상관관계",
          "strongCorrelations": "강함",
          "recentCorrelations": "최근 24시간",
          "recentTriggers": "자동 트리거 (24시간)"
        },
        "actions": {
          "viewDetails": "상세 보기",
          "goToAnomaly": "이상 탐지로 이동",
          "goToDrift": "드리프트 모니터링으로 이동",
          "refresh": "새로고침",
          "configure": "설정",
          "triggerNow": "지금 트리거",
          "saveConfig": "설정 저장"
        },
        "messages": {
          "configSaved": "설정이 저장되었습니다",
          "triggerStarted": "트리거가 시작되었습니다",
          "triggerCompleted": "트리거가 완료되었습니다",
          "triggerFailed": "트리거 실패",
          "noCorrelationsFound": "상관관계를 찾을 수 없습니다",
          "loadingCorrelations": "상관관계 로딩 중...",
          "errorLoadingCorrelations": "상관관계 로드 실패"
        },
        "empty": {
          "noCorrelations": "상관관계 없음",
          "noCorrelationsDesc": "동일 소스에서 이상 탐지와 드리프트 알림이 발생하면 상관관계가 표시됩니다",
          "noRelatedAlerts": "관련 알림 없음",
          "noRelatedAlertsDesc": "선택한 기간 내에 이 소스에 대한 연관된 알림이 없습니다",
          "noEvents": "자동 트리거 이벤트 없음",
          "noEventsDesc": "조건이 충족되면 자동 트리거 이벤트가 여기에 표시됩니다"
        },
        "time": {
          "seconds": "초",
          "minutes": "분",
          "hours": "시간",
          "ago": "전",
          "apart": "차이"
        }
      },
      "ja": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "zh": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "de": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "fr": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "es": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "pt": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "it": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "ru": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "ar": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "th": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "vi": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "id": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      },
      "tr": {
        "sections": {},
        "strength": {},
        "alertType": {},
        "labels": {},
        "config": {
          "thresholds": {}
        },
        "events": {
          "triggerType": {},
          "status": {}
        },
        "stats": {},
        "actions": {},
        "messages": {},
        "empty": {},
        "time": {}
      }
    }
  },
  "localIds": [
    "crossAlerts::local::src/content/cross-alerts.content.ts"
  ]
} as const;

File: src/truthound_dashboard/schemas/cross_alerts.py
"""Cross-alert correlation schemas.

This module defines schemas for cross-feature integration between
Anomaly Detection and Drift Monitoring alerts.

API Design: Direct Response Style
- Single resources return the resource directly
- List endpoints return PaginatedResponse with data, total, offset, limit
- Errors are handled via HTTPException
- Success is indicated by HTTP status codes (200, 201, 204)
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from .base import BaseSchema, IDMixin, TimestampMixin, PaginatedResponse


# =============================================================================
# Enums and Types
# =============================================================================

AlertType = Literal["anomaly", "drift"]
CorrelationStrength = Literal["strong", "moderate", "weak", "none"]
TriggerDirection = Literal["anomaly_to_drift", "drift_to_anomaly", "bidirectional"]


# =============================================================================
# Cross-Alert Correlation Schemas
# =============================================================================


class CorrelatedAlert(BaseSchema):
    """A single alert that is part of a correlation."""

    alert_id: str = Field(..., description="Alert ID")
    alert_type: AlertType = Field(..., description="Type of alert (anomaly or drift)")
    source_id: str = Field(..., description="Related source ID")
    source_name: str | None = Field(default=None, description="Source name for display")
    severity: str = Field(..., description="Alert severity level")
    message: str = Field(..., description="Alert message")
    created_at: datetime = Field(..., description="When the alert was created")

    # Type-specific fields
    anomaly_rate: float | None = Field(
        default=None,
        description="Anomaly rate (for anomaly alerts)",
    )
    anomaly_count: int | None = Field(
        default=None,
        description="Number of anomalies (for anomaly alerts)",
    )
    drift_percentage: float | None = Field(
        default=None,
        description="Drift percentage (for drift alerts)",
    )
    drifted_columns: list[str] | None = Field(
        default=None,
        description="Drifted columns (for drift alerts)",
    )


class CrossAlertCorrelation(IDMixin, TimestampMixin, BaseSchema):
    """A correlation between anomaly and drift alerts."""

    source_id: str = Field(..., description="Primary source ID")
    source_name: str | None = Field(default=None, description="Source name")

    # Correlation details
    correlation_strength: CorrelationStrength = Field(
        default="moderate",
        description="Strength of the correlation",
    )
    confidence_score: float = Field(
        default=0.0,
        ge=0.0,
        le=1.0,
        description="Confidence score of the correlation (0-1)",
    )
    time_delta_seconds: int = Field(
        default=0,
        description="Time difference between alerts in seconds",
    )

    # Related alerts
    anomaly_alert: CorrelatedAlert | None = Field(
        default=None,
        description="The anomaly alert in this correlation",
    )
    drift_alert: CorrelatedAlert | None = Field(
        default=None,
        description="The drift alert in this correlation",
    )

    # Analysis
    common_columns: list[str] = Field(
        default_factory=list,
        description="Columns affected by both anomaly and drift",
    )
    suggested_action: str | None = Field(
        default=None,
        description="Suggested remediation action",
    )
    notes: str | None = Field(
        default=None,
        description="Additional notes about the correlation",
    )


class CrossAlertCorrelationListResponse(PaginatedResponse[CrossAlertCorrelation]):
    """Paginated list of cross-alert correlations."""

    pass


class CorrelationSearchResult(BaseSchema):
    """Result of a correlation search for a specific source."""

    correlations: list[CrossAlertCorrelation] = Field(
        default_factory=list,
        description="Found correlations",
    )
    total: int = Field(default=0, description="Total correlations found")


# =============================================================================
# Auto-Trigger Configuration Schemas
# =============================================================================


class AutoTriggerThresholds(BaseSchema):
    """Thresholds for auto-triggering checks."""

    # Anomaly thresholds
    anomaly_rate_threshold: float = Field(
        default=0.1,
        ge=0.0,
        le=1.0,
        description="Minimum anomaly rate to trigger drift check",
    )
    anomaly_count_threshold: int = Field(
        default=10,
        ge=1,
        description="Minimum anomaly count to trigger drift check",
    )

    # Drift thresholds
    drift_percentage_threshold: float = Field(
        default=10.0,
        ge=0.0,
        le=100.0,
        description="Minimum drift percentage to trigger anomaly check",
    )
    drift_columns_threshold: int = Field(
        default=2,
        ge=1,
        description="Minimum drifted columns to trigger anomaly check",
    )


class AutoTriggerConfig(IDMixin, TimestampMixin, BaseSchema):
    """Configuration for auto-triggering cross-feature checks."""

    source_id: str | None = Field(
        default=None,
        description="Source ID (None for global config)",
    )

    # Enable/disable flags
    enabled: bool = Field(
        default=True,
        description="Whether auto-triggering is enabled",
    )
    trigger_drift_on_anomaly: bool = Field(
        default=True,
        description="Trigger drift check when anomaly detected",
    )
    trigger_anomaly_on_drift: bool = Field(
        default=True,
        description="Trigger anomaly check when drift detected",
    )

    # Thresholds
    thresholds: AutoTriggerThresholds = Field(
        default_factory=AutoTriggerThresholds,
        description="Trigger thresholds",
    )

    # Notification preferences
    notify_on_correlation: bool = Field(
        default=True,
        description="Send notification when correlation detected",
    )
    notification_channel_ids: list[str] | None = Field(
        default=None,
        description="Notification channel IDs",
    )

    # Cooldown to prevent alert storms
    cooldown_seconds: int = Field(
        default=300,
        ge=0,
        description="Minimum time between auto-triggered checks (seconds)",
    )
    last_anomaly_trigger_at: datetime | None = Field(
        default=None,
        description="Last time anomaly check was auto-triggered",
    )
    last_drift_trigger_at: datetime | None = Field(
        default=None,
        description="Last time drift check was auto-triggered",
    )


class AutoTriggerConfigCreate(BaseSchema):
    """Schema for creating auto-trigger configuration."""

    source_id: str | None = Field(default=None)
    enabled: bool = Field(default=True)
    trigger_drift_on_anomaly: bool = Field(default=True)
    trigger_anomaly_on_drift: bool = Field(default=True)
    thresholds: AutoTriggerThresholds | None = Field(default=None)
    notify_on_correlation: bool = Field(default=True)
    notification_channel_ids: list[str] | None = Field(default=None)
    cooldown_seconds: int = Field(default=300, ge=0)


class AutoTriggerConfigUpdate(BaseSchema):
    """Schema for updating auto-trigger configuration."""

    enabled: bool | None = Field(default=None)
    trigger_drift_on_anomaly: bool | None = Field(default=None)
    trigger_anomaly_on_drift: bool | None = Field(default=None)
    thresholds: AutoTriggerThresholds | None = Field(default=None)
    notify_on_correlation: bool | None = Field(default=None)
    notification_channel_ids: list[str] | None = Field(default=None)
    cooldown_seconds: int | None = Field(default=None, ge=0)


# =============================================================================
# Auto-Trigger Event Schemas
# =============================================================================


class AutoTriggerEvent(IDMixin, TimestampMixin, BaseSchema):
    """Record of an auto-triggered check."""

    source_id: str = Field(..., description="Source ID")
    trigger_type: TriggerDirection = Field(
        ...,
        description="Direction of the trigger",
    )

    # Trigger source
    trigger_alert_id: str = Field(..., description="ID of the alert that triggered this")
    trigger_alert_type: AlertType = Field(..., description="Type of triggering alert")

    # Result
    result_id: str | None = Field(
        default=None,
        description="ID of the resulting detection/comparison",
    )
    correlation_found: bool = Field(
        default=False,
        description="Whether a correlation was found",
    )
    correlation_id: str | None = Field(
        default=None,
        description="ID of the correlation if found",
    )

    # Status
    status: Literal["pending", "running", "completed", "failed", "skipped"] = Field(
        default="pending",
        description="Status of the auto-triggered check",
    )
    error_message: str | None = Field(
        default=None,
        description="Error message if failed",
    )
    skipped_reason: str | None = Field(
        default=None,
        description="Reason if skipped (e.g., cooldown)",
    )


class AutoTriggerEventListResponse(PaginatedResponse[AutoTriggerEvent]):
    """Paginated list of auto-trigger events."""

    pass


# =============================================================================
# Summary Schemas
# =============================================================================


class CrossAlertSummary(BaseSchema):
    """Summary of cross-alert correlations."""

    total_correlations: int = Field(
        default=0,
        description="Total number of correlations",
    )
    strong_correlations: int = Field(
        default=0,
        description="Number of strong correlations",
    )
    moderate_correlations: int = Field(
        default=0,
        description="Number of moderate correlations",
    )
    weak_correlations: int = Field(
        default=0,
        description="Number of weak correlations",
    )

    # Recent activity
    recent_correlations_24h: int = Field(
        default=0,
        description="Correlations in the last 24 hours",
    )
    recent_auto_triggers_24h: int = Field(
        default=0,
        description="Auto-triggers in the last 24 hours",
    )

    # Top affected sources
    top_affected_sources: list[dict] = Field(
        default_factory=list,
        description="Sources with most correlations",
    )

    # Auto-trigger stats
    auto_trigger_enabled: bool = Field(
        default=True,
        description="Whether auto-triggering is enabled globally",
    )
    anomaly_to_drift_triggers: int = Field(
        default=0,
        description="Count of anomaly-to-drift triggers",
    )
    drift_to_anomaly_triggers: int = Field(
        default=0,
        description="Count of drift-to-anomaly triggers",
    )
